src/Raven.Server/Commercial/LicenseManager.cs

```diff
@@ -121,40 +161,138 @@ public void Initialize(StorageEnvironment environment, TransactionContextPool co
             }
         }
 
-        public LicenseStatus GetLicenseStatus()
+        public async Task ReloadLicenseLimits()
         {
-            var lastLicenseIndex = _serverStore.LastLicenseIndex;
-            if (lastLicenseIndex > _lastSavedLicenseIndex)
+            try
             {
-                _lastSavedLicenseIndex = lastLicenseIndex;
-                ReloadLicense();
+                var licenseLimits = _serverStore.LoadLicenseLimits();
+                if (licenseLimits?.CoresByNode != null && 
+                    licenseLimits.CoresByNode.TryGetValue(_serverStore.NodeTag, out var cores))
+                {
+                    cores = Math.Min(cores, _licenseStatus.MaxCores);
+
+                    var process = Process.GetCurrentProcess();
+                    SetAffinity(process, cores);
+
+                    var ratio = (int)Math.Ceiling(_licenseStatus.MaxRamInGb / (double)_licenseStatus.MaxCores);
+                    var clusterSize = GetClusterSize();
+                    var maxWorkingSet = Math.Min(_licenseStatus.MaxRamInGb / (double)clusterSize, cores * ratio);
+                    SetMaxWorkingSet(process, Math.Max(1, maxWorkingSet));
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.Info("Failed to reload license limits", e);
             }
 
-            return _licenseStatus;
+            await RecalculateLicenseLimitsIfNeeded();
         }
 
-        private void ReloadLicense()
+        private int GetClusterSize()
         {
-            var license = _serverStore.LoadLicense();
-            if (license == null)
+            if (_serverStore.IsPassive())
+                return 1;
+
+            using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
             {
-                _licenseStatus.Attributes = null;
-                _licenseStatus.Error = false;
-                _licenseStatus.Message = null;
-                return;
+                return _serverStore.GetClusterTopology(context).AllNodes.Count;
             }
+        }
+
+        public async Task RecalculateLicenseLimitsIfNeeded(
+            string assignedNodeTag = null, int? assignedCores = null)
+        {
+            if (_serverStore.IsLeader() == false)
+                return;
 
             try
             {
-                _licenseStatus.Attributes = LicenseValidator.Validate(license, RSAParameters);
-                _licenseStatus.Error = false;
-                _licenseStatus.Message = null;
+                _recalculateLicenseLimitsSemsphore.Wait();
```

Comment:
Why wait like this? If we are held up there, we can give up and let the other one handle this, no?